### PR TITLE
Add Functionality For 'Embedding-Only' Training

### PR DESCRIPTION
**Pull request by nousr**
I thought I would just get the ball rolling on this front! Feel free to suggest changes or comment on errors in my code.

The main motivation for this pull request is outlined in #23.

After some discussion in the LAION discord, it was proposed to try a framework that makes the following major changes:
- `DiffusionPrior` can now be instantiated with `clip=None` to specify embedding-only training.
- `DiffusionPrior.forward()` now has two added parameters that default to None: `text_embed` and `image_embed`

I don't yet have access to the text embeddings from LAION2B, however, I tested on dummy data of the same shape and was able to run with no errors thrown--this example usage is included below.

### CODE REVIEW

File: dalle2_pytorch/dalle2_pytorch.py [hunk 1]
```diff
@@ -423,54 +423,77 @@ def forward(
...
-        # but let's just do it right
-
-        if exists(mask):
-            not_all_masked_out = mask.any(dim = -1)
-            mask = torch.cat((mask, rearrange(not_all_masked_out, 'b -> b 1')), dim = 1)
-
-        if exists(mask):
-            mask = F.pad(mask, (0, 2), value = True) # extend mask for text embedding, noised image embedding, time step embedding, and learned query
+        # grab some contextual information
+        batch, device = image_embed.shape[0], image_embed.device
```
Comment (nousr) [inline dalle2_pytorch/dalle2_pytorch.py:442]: pull out batch and device in the case of there being no text_encoding

File: dalle2_pytorch/dalle2_pytorch.py [hunk 2]
```diff
@@ -423,54 +423,77 @@ def forward(
...
 
+        # gather tokens for casual transformer
```
Comment (nousr) [inline dalle2_pytorch/dalle2_pytorch.py:444]: gathered all tokens to the same chunk of code; leaving the text_encoding which is now considered optional.

File: dalle2_pytorch/dalle2_pytorch.py [hunk 3]
```diff
@@ -423,54 +423,77 @@ def forward(
...
+        text_embed, image_embed = rearrange_many(
+            (text_embed, image_embed), 'b d -> b 1 d')
         time_embed = self.time_embeddings(diffusion_timesteps)
         time_embed = rearrange(time_embed, 'b d -> b 1 d')
+        learned_queries = repeat(self.learned_query, 'd -> b 1 d', b=batch)
+
+        # check for embedding only training
```
Comment (nousr) [inline dalle2_pytorch/dalle2_pytorch.py:451]: I opted to use a big if-statement block here to separate out the two possible training methods. I'm definitely open to more concise suggestions!

File: dalle2_pytorch/dalle2_pytorch.py [hunk 4]
```diff
@@ -679,15 +710,31 @@ def sample(self, text, num_samples_per_batch = 2):
...
-        check_shape(image, 'b c h w', h = img_size, w = img_size, c = self.channels)
+    def forward(self, text, image, text_embed=None, image_embed=None, *args, **kwargs):
+        # check if we are doing embedding-only training
+        if text is not None:
+            b, device, img_size, = image.shape[0], image.device, self.image_size
+            check_shape(image, 'b c h w', h=img_size,
+                        w=img_size, c=self.channels)
+            text_cond = self.get_text_cond(text)
+        else:
+            b, device = image_embed.shape[0], image_embed.device
```
Comment (nousr) [inline dalle2_pytorch/dalle2_pytorch.py:721]: The DiffusionPriorNetwork expects a dictionary. So I format the predefined text embedding in a dictionary with text_encodings:None to make it possible to detect the embedding-only training once inside the DiffusionPriorNetwork.

File: dalle2_pytorch/dalle2_pytorch.py [hunk 5]
```diff
@@ -423,54 +423,77 @@ def forward(
...
+            ), dim=-2)
+
+        # TODO: someone should double check this block of code below
+        else:
+            # during embedding only, just drop text_encodings entirely
+            tokens = torch.cat((
+                text_embed,
+                time_embed,
+                learned_queries
+            ), dim=-2)
```
Comment (nousr) [inline dalle2_pytorch/dalle2_pytorch.py:488]: I'd like someone to check this part (above) and verify that we can simply use a mask of all ones in the shape of `(batch, tokens.shape[1])`.

My understanding is that the mask is only used for the classifier-free guidance training (text_encoding). But I could be wrong!

**Review by nousr — COMMENTED** [review]
I went through and added some comments on the changes I made to hopefully give some context for the problem and thought process for the changes.

File: dalle2_pytorch/dalle2_pytorch.py [hunk 6]
```diff
@@ -423,54 +423,77 @@ def forward(
...
 
+        # gather tokens for casual transformer
+        text_embed, image_embed = rearrange_many(
+            (text_embed, image_embed), 'b d -> b 1 d')
         time_embed = self.time_embeddings(diffusion_timesteps)
         time_embed = rearrange(time_embed, 'b d -> b 1 d')
+        learned_queries = repeat(self.learned_query, 'd -> b 1 d', b=batch)
+
+        # check for embedding only training
+        if text_encodings is not None:
```
Comment (nousr) [inline dalle2_pytorch/dalle2_pytorch.py:452]: I opted to use a big if-statement block here to separate out the two possible training methods. I'm definitely open to more concise suggestions!

File: dalle2_pytorch/dalle2_pytorch.py [hunk 7]
```diff
@@ -423,54 +423,77 @@ def forward(
...
-            mask = F.pad(mask, (0, 2), value = True) # extend mask for text embedding, noised image embedding, time step embedding, and learned query
+        # grab some contextual information
+        batch, device = image_embed.shape[0], image_embed.device
 
+        # gather tokens for casual transformer
+        text_embed, image_embed = rearrange_many(
+            (text_embed, image_embed), 'b d -> b 1 d')
         time_embed = self.time_embeddings(diffusion_timesteps)
         time_embed = rearrange(time_embed, 'b d -> b 1 d')
+        learned_queries = repeat(self.learned_query, 'd -> b 1 d', b=batch)
```
Comment (nousr) [inline dalle2_pytorch/dalle2_pytorch.py:449]: gathered all tokens to the same chunk of code; leaving the text_encoding which is now considered optional.

**Review by nousr — COMMENTED** [review]
updated some comments

File: dalle2_pytorch/dalle2_pytorch.py (hunk 5 above)
Comment (lucidrains) [reply dalle2_pytorch/dalle2_pytorch.py:488]: yup this looks good! 💯

File: dalle2_pytorch/dalle2_pytorch.py (hunk 6 above)
Comment (lucidrains) [reply dalle2_pytorch/dalle2_pytorch.py:452]: yup, that's fine, another way would be to set text encodings to an empty tensor with the dimension at which to be concatenated to be 0

**Comment by nousr** [comment]
> I tested on dummy data of the same shape and was able to run with no errors thrown--this example usage is included below.

```python

"""
Embedding-Only Training: Proof of Concept
"""

import torch
from dalle2_pytorch import DiffusionPriorNetwork, DiffusionPrior

# grab our models
prior_network = DiffusionPriorNetwork(
    dim=768,
    depth=6,
    dim_head=64,
    heads=8
).cuda()

# specify clip=None to signal for "embedding-only" training
diffusion_prior = DiffusionPrior(
    net=prior_network,
    clip=None,
    timesteps=64,
    cond_drop_prob=1
).cuda()

# LAION embeddings are (b, 768)
txt_embed = torch.normal(0, 1, (10, 768)).cuda()
img_embed = torch.normal(0, 1, (10, 768)).cuda()

# quick boilerplate 'training' to sanity check
loss = 0.0
EPOCHS = 10
optimizer = torch.optim.Adam(diffusion_prior.parameters())

for e in range(EPOCHS):

    optimizer.zero_grad()

    loss = diffusion_prior(text=None, image=None,
                           text_embed=txt_embed, image_embed=img_embed)
    loss.backward()
    loss = loss.item()
    optimizer.step()

    print(f"Epoch {e} | Loss: {loss}")

print(f"Test Completed!")```

File: dalle2_pytorch/dalle2_pytorch.py (hunk 4 above)
Comment (lucidrains) [reply dalle2_pytorch/dalle2_pytorch.py:721]: yea, so here i'd actually just do `def forward(self, text = None, image = None, text_embed = None, image_embed = None)`

and later on do asserts to make sure one or the other is passed in `assert exists(text) ^ exists(text_embed)` and `assert exists(image) ^ exists(image_embed)` and handle them both in separate conditionals, prepping the kwargs for the `p_losses` function below

**Comment by lucidrains** [comment]
love it! let's get this merged tomorrow morning (i'm off to 💤 )

**Comment by lucidrains** [comment]
@nousr hey, i'm really sorry for doing this, but i think i'm going to hijack this and just knock out the code in the next hour or two. because you plan on training the embeddings alone (and it seems like you are also omitting text encodings), i will probably have the researcher specify upfront whether text encodings will be used. this will also change the logic for how i do the classifier free guidance (which is off the mask of the text encodings). and the third complication would be to account for all this when sampling

valiant efforts otherwise! glad to have someone else looking at the code with me!

**Comment by nousr** [comment]
@lucidrains No worries! Thanks for taking the time to give some feedback. I look forward to the updates 🚀

**Comment by lucidrains** [comment]
@nousr ok, let me know if this works for you! https://github.com/lucidrains/DALLE2-pytorch/tree/0.0.48#training-on-preprocessed-clip-embeddings (open up a new github issue if it does not)
